src/examples/specgan_example/specgan_example.py

```diff
@@ -0,0 +1,274 @@
+        ), "Spectogram dimensions is not divisible by 4"
+
+        # Create preprocessing pipeline for shuffling and batching
+        return preprocess.pipeline(
+            [
+                preprocess.set_channels(1),
+                preprocess.shuffle(self.hparams["buffer_size"]),
+                preprocess.batch(self.hparams["batch_size"]),
+                preprocess.prefetch(),
+            ]
+        )(dataset)
+
+    # This runs at the start of every epoch
+    def on_epoch_start(self, epoch, step):
+        self.gen_loss_avg = tf.keras.metrics.Mean()
+        self.disc_loss_avg = tf.keras.metrics.Mean()
+
+    # This runs at every step in the training (for each batch in dataset)
+    def on_step(self, step, stats):
+        gen_loss, disc_loss = stats
+        self.gen_loss_avg(gen_loss)
+        self.disc_loss_avg(disc_loss)
+        if step % 100 == 0:
+            print(
+                f"Step: {step}, Gen Loss: {self.gen_loss_avg.result()}, Disc Loss: {self.disc_loss_avg.result()}"
+            )
+
+    # This runs at the end of every epoch and is used to display metrics
+    def on_epoch_complete(self, epoch, step, duration):
+        # display.clear_output(wait=True)
+        print(
+            f"Epoch: {epoch}, Step: {step}, Gen Loss: {self.gen_loss_avg.result()}, Disc Loss: {self.disc_loss_avg.result()}, Duration: {duration} s"
+        )
+        self.generate_and_save_images_epoch(epoch, step)
+
+    def generate_and_save_images_epoch(self, epoch, step):
+        generated = self.generator(self.seed, training=False)
+
+        fig = plt.figure(figsize=(4, 4))
+
+        for i in range(generated.shape[0]):
+            plt.subplot(4, 4, i + 1)
+            plt.imshow(generated[i, :, :, 0])
+            plt.axis("off")
+
+        if "save_images" in self.hparams and self.hparams["save_images"]:
+            if self.image_save_dir is None:
+                raise Exception(
+                    "Could not save image, no save_dir was specified in hparams."
+                )
+            plt.savefig(
+                os.path.join(
+                    self.image_save_dir,
+                    "image_at_epoch_{:04d}_step_{}.png".format(epoch, step),
+                )
+            )
+
+        if self.hparams["plot"]:
+            plt.show()
+
+    def sample_sound(self, seed, pipeline):
+        generated = self.generator(seed, training=False)
+
+        print("Generated melspec samples:")
+        fig = plt.figure(figsize=(4, 4))
+
+        for i in range(generated.shape[0]):
+            plt.subplot(4, 4, i + 1)
+            plt.imshow(generated[i, :, :, 0])
+            plt.axis("off")
+
+        if self.hparams["plot"]:
+            plt.show()
+
+        return pipeline(tf.unstack(generated))
+
+
+if __name__ == "__main__":
+    # Load hyperparams from yaml file
+    defaults = load_hparams("hparams.yaml")
+
+    parser = argparse.ArgumentParser(description="Start training of the gan_example")
+    parser.add_argument(
```

Comment:
I usually like to avoid going over 80 characters, but in this case I think I agree that the arguments should be on one line since this code isn't too important to be readable. 
We might even want to consider moving the parsing stuff to another `util` function.

I think the argument parser has support for short-hands as well.

Reply:
I usually prefer a strategy of "optimize legibility" than "never ever exceed 80". And yeah, it's just `add_argument("-p", "--plot", ...)` for shorthands. I'm all for splitting the options into a separate file, but not all models have the same hyperparameters and stuff.

Reply:
Again, python black enforces it but they do have a very strict guideline regarding long lines.